pylint/utils/utils.py

```diff
@@ -91,18 +91,18 @@ def tokenize_module(module):
         return list(tokenize.tokenize(readline))
 
 
-def _basename_in_blacklist_re(base_name, black_list_re):
+def _is_in_blacklist_re(path, black_list_re):
     """Determines if the basename is matched in a regex blacklist
 
-    :param str base_name: The basename of the file
+    :param str path: The file path
     :param list black_list_re: A collection of regex patterns to match against.
         Successful matches are blacklisted.
 
     :returns: `True` if the basename is blacklisted, `False` otherwise.
     :rtype: bool
     """
     for file_pattern in black_list_re:
-        if file_pattern.match(base_name):
+        if file_pattern.match(path):
```

Comment:
I believe this is a backwards incompatible change. Users that currently rely on `ignore-patterns` will suddenly see warnings from files they previously ignored. Given a full match might be too much, we might use `.search` here instead. 
We should add a test for that to make sure what patterns that assumed this works on the basename will still filter out the corresponding files.

Reply:
Hi @PCManticore Thank you for your constructive comment. `.search` would indeed decrease the number of cases where backward-incompatibility occurs, but won't completely eliminate them, particularly when the tree structure presents repeating patterns, e.g., with `ignore-patterns=foo`:
```
whatever/foo/bar.py
whatever/foo/foo-a.py
whatever/foo/foo-b.py
```
The current implementation would lint `bar.py`, while this PR with `.match` or `.search` would ignore it

I'm afraid that to be fully backward compatible without anyone having to rewrite its `ignore-patterns` configuration directive, we'd have to implement a new one, such as `ignore-path-patterns`, what I can do.

What do you think ?

Reply:
Hello, kind reminder on this old PR.
Personally I find new separate option safer and more flexible.

Reply:
I can update it, but would prefer to have @PCManticore thoughts before doing it

Reply:
@bernardn Sorry for letting this fall through the cracks. Here are my thoughts:
* I think we should be backwards compatible to mess as much with users configuration as possible. Let's go with a new `ignore-paths` option, which will support regexes.
* This will bring the choice for ignoring directories to three: `ignore-paths`, `ignore` and `ignore-patterns`. Once we have the first, we can start deprecating the other two, so we could remove them in pylint 3.0. We'll emit deprecation warnings when detecting their usage.
* Let's throw an error if these three are used together. e.g. detect if `ignore-paths` is used with `ignore-patterns` and simply throw a new error from `pylint.exceptions` that this behaviour is not supported.

Let me know if that makes sense to you as well.

Reply:
@PCManticore This makes perfect sense to me as well, thank you! I'm going to implement this in the upcoming days.